modules/apps/portal-workflow/portal-workflow-kaleo-runtime-impl/src/main/java/com/liferay/portal/workflow/kaleo/runtime/internal/util/RoleNotificationRecipientBuilderUtil.java

```diff
@@ -0,0 +1,274 @@
+			}
+		}
+
+		return groupIds;
+	}
+
+	private static List<User> _getRoleUsers(
+			Role role, ExecutionContext executionContext)
+		throws Exception {
+
+		long roleId = role.getRoleId();
+
+		UserLocalService userLocalService = _userLocalServiceSnapshot.get();
+
+		if (role.getType() == RoleConstants.TYPE_REGULAR) {
+			return userLocalService.getInheritedRoleUsers(
+				roleId, QueryUtil.ALL_POS, QueryUtil.ALL_POS, null);
+		}
+
+		KaleoInstanceToken kaleoInstanceToken =
+			executionContext.getKaleoInstanceToken();
+
+		List<Long> groupIds = _getGroupIds(
+			kaleoInstanceToken.getGroupId(), role);
+
+		List<User> users = new ArrayList<>();
+
+		for (Long groupId : groupIds) {
+			UserGroupRoleLocalService userGroupRoleLocalService =
+				_userGroupRoleLocalServiceSnapshot.get();
+
+			List<UserGroupRole> userGroupRoles =
+				userGroupRoleLocalService.getUserGroupRolesByGroupAndRole(
+					groupId, roleId);
+
+			for (UserGroupRole userGroupRole : userGroupRoles) {
+				users.add(userGroupRole.getUser());
+			}
+
+			UserGroupGroupRoleLocalService userGroupGroupRoleLocalService =
+				_userGroupGroupRoleLocalServiceSnapshot.get();
+
+			List<UserGroupGroupRole> userGroupGroupRoles =
+				userGroupGroupRoleLocalService.
+					getUserGroupGroupRolesByGroupAndRole(groupId, roleId);
+
+			for (UserGroupGroupRole userGroupGroupRole : userGroupGroupRoles) {
+				users.addAll(
+					userLocalService.getUserGroupUsers(
+						userGroupGroupRole.getUserGroupId()));
+			}
+
+			if (Objects.equals(
+					role.getName(), DepotRolesConstants.ASSET_LIBRARY_MEMBER) ||
+				Objects.equals(role.getName(), RoleConstants.SITE_MEMBER)) {
+
+				users.addAll(
+					userLocalService.getGroupUsers(
+						groupId, WorkflowConstants.STATUS_APPROVED, null));
+			}
+
+			if (Objects.equals(
+					role.getName(), RoleConstants.ORGANIZATION_USER)) {
+
+				GroupLocalService groupLocalService =
+					_groupLocalServiceSnapshot.get();
+
+				Group group = groupLocalService.getGroup(groupId);
+
+				if (group.isOrganization()) {
+					long organizationId = group.getClassPK();
+
+					users.addAll(
+						userLocalService.getOrganizationUsers(organizationId));
+				}
+			}
+		}
+
+		return users;
+	}
+
+	private static boolean _isValidGroup(Group group, Role role)
+		throws Exception {
+
+		if ((group != null) && group.isDepot() &&
+			(role.getType() == RoleConstants.TYPE_DEPOT)) {
+
+			return true;
+		}
+		else if ((group != null) && group.isOrganization() &&
+				 (role.getType() == RoleConstants.TYPE_ORGANIZATION)) {
+
+			return true;
+		}
+		else if ((group != null) && group.isSite() &&
+				 (role.getType() == RoleConstants.TYPE_SITE)) {
+
+			return true;
+		}
+
+		for (GroupAwareRoleValidator groupAwareRoleValidator :
+				_serviceTrackerList) {
+
+			if (groupAwareRoleValidator.isValidGroup(group, role)) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	private static final Snapshot<GroupLocalService>
+		_groupLocalServiceSnapshot = new Snapshot<>(
```

Comment:
@RenanUmV We need to avoid using `Snapshot` to replace any hard/normal reference.